libs/kafka/consumer.go

```diff
@@ -0,0 +1,66 @@
+package kafka
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/brave-intl/bat-go/libs/logging"
+	"github.com/getsentry/sentry-go"
+	kafkago "github.com/segmentio/kafka-go"
+)
+
+type Handler interface {
+	Handle(ctx context.Context, message kafkago.Message) error
+}
+
+// ErrorHandler defines
+type ErrorHandler interface {
+	Handle(ctx context.Context, message kafkago.Message, errorMessage error) error
+}
+
+// TODO: Commit many messages off cache.
+// TODO: Return and flush offset is fail deadletter
+// TODO: Add support multithreading handler
+
+// Consume implements consumer loop.
+func Consume(ctx context.Context, reader Consumer, handler Handler, errorHandler ErrorHandler) (bool, error) {
+	logger := logging.Logger(ctx, "kafka consumer")
+	logger.Info().Msg("starting consumer")
+
+	for {
+		select {
+		case <-ctx.Done():
+			return true, ctx.Err()
+		default:
+			message, err := reader.FetchMessage(ctx)
+			if err != nil {
+				return true, fmt.Errorf("error fetching message key %s partition %d offset %d: %w",
+					string(message.Key), message.Partition, message.Offset, err)
+			}
+
+			err = handler.Handle(ctx, message)
```

Comment:
since there is no return / continue in this branch - in the case that there was some sort of temporary processing error then we will dump the message on the floor and never process it

Reply:
If there is an error processing the message it should will be handed off to the errorHandler and dlq where it can be retried off the main thread, if we cant get to the dlq then we log and can manually rerun. There is retry built into the writers. There is a `TODO Return and flush offset is fail deadletter` which would also need a refactor to allow the consumers to self heal. I can address that now

Reply:
addressing now

Reply:
The idea is if the handler fails the message is handled off to the dlq (whose writer will try a bunch of times) and if that fails we log the message and would need to rerun the message manually. The intention was to keep this really simple, and not get too far into the Kafka refactor (which is in the other ticket). I have added a basic retry mechanism in the service so if we fail to dlq then we will restart the consumer after a backoff.